// docmd-search/dist/index.js
var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __esm = (fn, res, err) => function __init() {
  if (err) throw err[0];
  try {
    return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
  } catch (e) {
    throw err = [e], e;
  }
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/gpu.ts
var gpu_exports = {};
__export(gpu_exports, {
  detectGPUBackend: () => detectGPUBackend,
  formatGPUInfo: () => formatGPUInfo,
  getExecutionProviders: () => getExecutionProviders
});
import { platform } from "os";
async function detectGPUBackend() {
  const currentPlatform = platform();
  if (typeof window !== "undefined") {
    if ("gpu" in navigator) {
      try {
        const adapter = await navigator.gpu.requestAdapter();
        if (adapter) {
          return {
            backend: "webgpu",
            name: "WebGPU",
            hasGPU: true,
            details: "Browser GPU acceleration"
          };
        }
      } catch {
      }
    }
    return {
      backend: "cpu",
      name: "CPU (WebAssembly)",
      hasGPU: false,
      details: "WebGPU not available"
    };
  }
  try {
    const ort = await import("onnxruntime-node");
    if (currentPlatform === "darwin") {
      const hasMPS = await checkMPSAvailability(ort);
      if (hasMPS) {
        return {
          backend: "mps",
          name: "Metal Performance Shaders (MPS)",
          hasGPU: true,
          details: "Apple Silicon GPU + Neural Engine"
        };
      }
    }
    if (currentPlatform === "win32" || currentPlatform === "linux") {
      const hasCUDA = await checkCUDAAvailability(ort);
      if (hasCUDA) {
        return {
          backend: "cuda",
          name: "CUDA",
          hasGPU: true,
          details: "NVIDIA GPU acceleration"
        };
      }
    }
    if (currentPlatform === "win32") {
      const hasDirectML = await checkDirectMLAvailability(ort);
      if (hasDirectML) {
        return {
          backend: "directml",
          name: "DirectML",
          hasGPU: true,
          details: "Windows GPU acceleration"
        };
      }
    }
  } catch (err) {
  }
  return {
    backend: "cpu",
    name: "CPU (WebAssembly)",
    hasGPU: currentPlatform === "darwin" ? true : false,
    // Mac has excellent CPU performance
    details: currentPlatform === "darwin" ? "Apple Silicon CPU (optimized)" : "Multi-threaded CPU inference"
  };
}
async function checkMPSAvailability(ort) {
  try {
    const providers = ort.listSupportedExecutionProviders?.() || [];
    return providers.includes("mps") || providers.includes("MPS");
  } catch {
    return platform() === "darwin";
  }
}
async function checkCUDAAvailability(ort) {
  try {
    const providers = ort.listSupportedExecutionProviders?.() || [];
    return providers.includes("cuda") || providers.includes("CUDA");
  } catch {
    return false;
  }
}
async function checkDirectMLAvailability(ort) {
  try {
    const providers = ort.listSupportedExecutionProviders?.() || [];
    return providers.includes("dml") || providers.includes("DirectML");
  } catch {
    return false;
  }
}
function getExecutionProviders(gpu) {
  switch (gpu.backend) {
    case "mps":
      return ["mps", "cpu"];
    case "cuda":
      return ["cuda", "cpu"];
    case "directml":
      return ["dml", "cpu"];
    case "webgpu":
      return ["webgpu", "cpu"];
    default:
      return ["cpu"];
  }
}
function formatGPUInfo(gpu) {
  if (gpu.hasGPU) {
    return `\u26A1 ${gpu.name}`;
  }
  return `\u25C7 ${gpu.name}`;
}
var init_gpu = __esm({
  "src/gpu.ts"() {
    "use strict";
  }
});

// src/parallel-embed.ts
var parallel_embed_exports = {};
__export(parallel_embed_exports, {
  ParallelEmbedder: () => ParallelEmbedder,
  embedParallel: () => embedParallel
});
import { Worker } from "worker_threads";
import * as os from "os";
import * as path from "path";
import { fileURLToPath } from "url";
async function embedParallel(texts, options) {
  const batchSize = 64;
  const batches = [];
  for (let i = 0; i < texts.length; i += batchSize) {
    batches.push(texts.slice(i, i + batchSize));
  }
  const embedder = new ParallelEmbedder(options);
  try {
    const results = await embedder.embedAll(batches);
    const allEmbeddings = [];
    for (const batchEmbeddings of results) {
      for (const embedding of batchEmbeddings) {
        allEmbeddings.push(new Float32Array(embedding));
      }
    }
    return allEmbeddings;
  } finally {
    await embedder.terminate();
  }
}
var __filename, __dirname, ParallelEmbedder;
var init_parallel_embed = __esm({
  "src/parallel-embed.ts"() {
    "use strict";
    __filename = fileURLToPath(import.meta.url);
    __dirname = path.dirname(__filename);
    ParallelEmbedder = class {
      constructor(options) {
        this.options = options;
        this.numWorkers = options.numWorkers || Math.max(1, os.cpus().length - 1);
        this.onProgress = options.onProgress;
        this.initPromise = this.initialize();
      }
      options;
      workers = [];
      freeWorkers = [];
      taskQueue = [];
      pendingResults = /* @__PURE__ */ new Map();
      totalTasks = 0;
      completedTasks = 0;
      onProgress;
      isInitialized = false;
      initPromise;
      numWorkers;
      async initialize() {
        const workerScript = path.join(__dirname, "embedding-worker.js");
        const readyPromises = [];
        for (let i = 0; i < this.numWorkers; i++) {
          const worker = new Worker(workerScript, {
            workerData: {
              modelId: this.options.modelId
            }
          });
          const readyPromise = new Promise((resolve2, reject) => {
            const timeout = setTimeout(() => {
              reject(new Error(`Worker ${i} failed to initialize within 60 seconds`));
            }, 6e4);
            worker.on("message", (message) => {
              if (message.type === "ready") {
                clearTimeout(timeout);
                this.freeWorkers.push(worker);
                resolve2();
              } else if (message.type === "result") {
                this.pendingResults.set(message.batchId, message.embeddings);
                this.completedTasks++;
                this.onProgress?.(this.completedTasks, this.totalTasks);
                this.freeWorkers.push(worker);
                this.processQueue();
              } else if (message.type === "error") {
                console.error("Worker error:", message.error);
                this.freeWorkers.push(worker);
                this.processQueue();
              }
            });
            worker.on("error", (err) => {
              clearTimeout(timeout);
              console.error("Worker thread error:", err);
              this.freeWorkers.push(worker);
              this.processQueue();
              reject(err);
            });
          });
          readyPromises.push(readyPromise);
          this.workers.push(worker);
        }
        await Promise.all(readyPromises);
        this.isInitialized = true;
      }
      async embedAll(batches) {
        await this.initPromise;
        return new Promise((resolve2, reject) => {
          this.totalTasks = batches.length;
          this.completedTasks = 0;
          this.pendingResults.clear();
          let completed = 0;
          const results = new Array(batches.length);
          batches.forEach((batch, batchId) => {
            this.taskQueue.push({
              batchId,
              batch,
              resolve: (embeddings) => {
                results[batchId] = embeddings;
                completed++;
                if (completed === batches.length) {
                  resolve2(results);
                }
              },
              reject
            });
          });
          this.processQueue();
        });
      }
      processQueue() {
        while (this.taskQueue.length > 0 && this.freeWorkers.length > 0) {
          const worker = this.freeWorkers.pop();
          const task = this.taskQueue.shift();
          const handler = (message) => {
            if (message.type === "result" && message.batchId === task.batchId) {
              worker.off("message", handler);
              task.resolve(message.embeddings);
            } else if (message.type === "error" && message.batchId === task.batchId) {
              worker.off("message", handler);
              task.reject(new Error(message.error));
            }
          };
          worker.on("message", handler);
          worker.postMessage({ type: "embed", batchId: task.batchId, batch: task.batch });
        }
      }
      async terminate() {
        await Promise.all(this.workers.map((w) => w.terminate()));
        this.workers = [];
        this.freeWorkers = [];
      }
    };
  }
});

// src/index-io.ts
var index_io_exports = {};
__export(index_io_exports, {
  buildNavTree: () => buildNavTree,
  compressVectors: () => compressVectors,
  createEmptyManifest: () => createEmptyManifest,
  createSearchIndex: () => createSearchIndex,
  decompressVectors: () => decompressVectors,
  getCompressionType: () => getCompressionType,
  hasSearchableIndex: () => hasSearchableIndex,
  loadAllBatches: () => loadAllBatches,
  loadBatch: () => loadBatch,
  loadManifest: () => loadManifest,
  loadSearchIndex: () => loadSearchIndex,
  saveBatch: () => saveBatch,
  saveManifest: () => saveManifest,
  saveNavigation: () => saveNavigation,
  updateManifest: () => updateManifest
});
import { writeFile as writeFile2, readFile as readFile3, mkdir as mkdir2 } from "node:fs/promises";
import { existsSync } from "node:fs";
import { join as join4 } from "node:path";
function getCompressionType(totalChunks) {
  if (totalChunks >= 2e3) return "pq";
  if (totalChunks >= 100) return "ternary";
  return "none";
}
function ternaryCompress(vectors, dimensions) {
  const results = [];
  for (const vec of vectors) {
    const bytesNeeded = Math.ceil(dimensions / 4);
    const packed = Buffer.alloc(bytesNeeded);
    for (let i = 0; i < dimensions; i++) {
      const val = vec[i];
      const ternary = val < -42 ? 0 : val > 42 ? 2 : 1;
      const byteIdx = Math.floor(i / 4);
      const bitOffset = i % 4 * 2;
      packed[byteIdx] |= ternary << bitOffset;
    }
    results.push(packed);
  }
  return Buffer.concat(results);
}
function ternaryDecompress(buffer, vectorCount, dimensions) {
  const bytesPerVector = Math.ceil(dimensions / 4);
  const vectors = [];
  for (let v = 0; v < vectorCount; v++) {
    const vec = new Int8Array(dimensions);
    const offset = v * bytesPerVector;
    for (let i = 0; i < dimensions; i++) {
      const byteIdx = offset + Math.floor(i / 4);
      const bitOffset = i % 4 * 2;
      const ternary = buffer[byteIdx] >> bitOffset & 3;
      vec[i] = ternary === 0 ? -127 : ternary === 2 ? 127 : 0;
    }
    vectors.push(vec);
  }
  return vectors;
}
function compressVectors(vectors, dimensions, type) {
  if (type === "none") {
    const buffer = Buffer.alloc(vectors.length * dimensions);
    for (let i = 0; i < vectors.length; i++) {
      buffer.set(vectors[i], i * dimensions);
    }
    return buffer;
  }
  return ternaryCompress(vectors, dimensions);
}
function decompressVectors(buffer, vectorCount, dimensions, type) {
  if (type === "none") {
    const vectors = [];
    for (let i = 0; i < vectorCount; i++) {
      const start = i * dimensions;
      vectors.push(new Int8Array(buffer.buffer, buffer.byteOffset + start, dimensions));
    }
    return vectors;
  }
  return ternaryDecompress(buffer, vectorCount, dimensions);
}
function padBatchId(id) {
  return String(id).padStart(3, "0");
}
async function saveBatch(outDir, batchId, chunks, vectors, dimensions, compression) {
  const batchDir = join4(outDir, "batches");
  await mkdir2(batchDir, { recursive: true });
  const prefix = padBatchId(batchId);
  const comp = compression ?? "none";
  const meta = {
    batchId,
    chunks,
    dimensions,
    compression: comp,
    vectorCount: vectors.length
  };
  await writeFile2(join4(batchDir, `${prefix}.json`), JSON.stringify(meta), "utf-8");
  const binData = compressVectors(vectors, dimensions, comp);
  await writeFile2(join4(batchDir, `${prefix}.bin`), binData);
}
async function loadBatch(dir, batchId) {
  const prefix = padBatchId(batchId);
  const batchDir = join4(dir, "batches");
  const meta = JSON.parse(
    await readFile3(join4(batchDir, `${prefix}.json`), "utf-8")
  );
  const bin = await readFile3(join4(batchDir, `${prefix}.bin`));
  const buffer = Buffer.from(bin.buffer, bin.byteOffset, bin.byteLength);
  const vectors = decompressVectors(buffer, meta.vectorCount, meta.dimensions, meta.compression);
  return { chunks: meta.chunks, vectors };
}
async function loadManifest(dir) {
  try {
    const raw = await readFile3(join4(dir, MANIFEST_FILE), "utf-8");
    return JSON.parse(raw);
  } catch {
    return null;
  }
}
async function saveManifest(dir, manifest) {
  await mkdir2(dir, { recursive: true });
  await writeFile2(join4(dir, MANIFEST_FILE), JSON.stringify(manifest, null, 2), "utf-8");
}
async function updateManifest(dir, updates) {
  const existing = await loadManifest(dir);
  const manifest = {
    ...existing ?? createEmptyManifest("", 0),
    ...updates
  };
  await saveManifest(dir, manifest);
  return manifest;
}
function createEmptyManifest(model, dimensions) {
  return {
    version: 1,
    model,
    dimensions,
    status: "indexing",
    totalFiles: 0,
    indexedFiles: 0,
    totalChunks: 0,
    batchCount: 0,
    files: {}
  };
}
function buildNavTree(files) {
  const root = [];
  const dirMap = /* @__PURE__ */ new Map();
  const sorted = [...files].sort((a, b) => a.path.localeCompare(b.path));
  for (const file of sorted) {
    const parts = file.path.split("/");
    const fileName = parts.pop();
    let currentLevel = root;
    let currentPath = "";
    for (const part of parts) {
      currentPath = currentPath ? `${currentPath}/${part}` : part;
      if (!dirMap.has(currentPath)) {
        const dirNode = {
          title: formatTitle(part),
          path: currentPath,
          children: []
        };
        currentLevel.push(dirNode);
        dirMap.set(currentPath, dirNode);
      }
      currentLevel = dirMap.get(currentPath).children;
    }
    currentLevel.push({
      title: file.title || formatTitle(fileName.replace(/\.\w+$/, "")),
      path: file.path
    });
  }
  return root;
}
function formatTitle(slug) {
  return slug.replace(/[-_]/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
}
async function saveNavigation(dir, tree) {
  await mkdir2(dir, { recursive: true });
  const nav = { tree };
  await writeFile2(join4(dir, "navigation.json"), JSON.stringify(nav, null, 2), "utf-8");
}
function hasSearchableIndex(dir) {
  return existsSync(join4(dir, "batches", "000.json"));
}
async function loadAllBatches(dir) {
  const manifest = await loadManifest(dir);
  if (!manifest) {
    throw new Error(`No index manifest found in ${dir}`);
  }
  const allChunks = [];
  const allVectors = [];
  for (let i = 0; i < manifest.batchCount; i++) {
    const { chunks, vectors } = await loadBatch(dir, i);
    allChunks.push(...chunks);
    allVectors.push(...vectors);
  }
  return {
    version: manifest.version,
    model: manifest.model,
    dimensions: manifest.dimensions,
    chunks: allChunks,
    vectors: allVectors
  };
}
async function createSearchIndex(index, outDir, options) {
  await mkdir2(outDir, { recursive: true });
  const compression = getCompressionType(index.chunks.length);
  await saveBatch(outDir, 0, index.chunks, index.vectors, index.dimensions, compression);
  const manifest = createEmptyManifest(index.model, index.dimensions);
  manifest.status = "complete";
  manifest.totalChunks = index.chunks.length;
  manifest.batchCount = 1;
  await saveManifest(outDir, manifest);
  if (!options?.silent) {
    const metaSize = (JSON.stringify(index.chunks).length / 1024).toFixed(1);
    const vecSize = (index.vectors.length * index.dimensions / 1024).toFixed(1);
    console.log(`  Index written \u2192 ${outDir}`);
    console.log(`    meta: ${metaSize} KB \xB7 vectors: ${vecSize} KB \xB7 compression: ${compression}`);
  }
}
async function loadSearchIndex(dir) {
  const manifest = await loadManifest(dir);
  if (manifest) {
    return loadAllBatches(dir);
  }
  const META_FILE = "search-index.json";
  const INDEX_FILE = "search-index.bin";
  const meta = JSON.parse(await readFile3(join4(dir, META_FILE), "utf-8"));
  const bin = await readFile3(join4(dir, INDEX_FILE));
  const flat = new Int8Array(bin.buffer, bin.byteOffset, bin.byteLength);
  const vectors = [];
  for (let i = 0; i < meta.chunks.length; i++) {
    vectors.push(flat.slice(i * meta.dimensions, (i + 1) * meta.dimensions));
  }
  return { ...meta, vectors };
}
var MANIFEST_FILE;
var init_index_io = __esm({
  "src/index-io.ts"() {
    "use strict";
    MANIFEST_FILE = "manifest.json";
  }
});

// src/indexer/index.ts
import { stat as stat3 } from "node:fs/promises";
import { relative as relative3, join as join6, isAbsolute } from "node:path";

// src/indexer/crawl.ts
import { readdir } from "node:fs/promises";
import { join, relative } from "node:path";
async function crawl(rootDir, include, exclude) {
  const results = [];
  async function walk(dir) {
    const entries = await readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = join(dir, entry.name);
      const rel = relative(rootDir, fullPath);
      if (exclude.some((p) => matchSimpleGlob(rel, p))) continue;
      if (entry.isDirectory()) {
        await walk(fullPath);
      } else if (include.some((p) => matchSimpleGlob(rel, p))) {
        results.push(fullPath);
      }
    }
  }
  await walk(rootDir);
  return results;
}
function matchSimpleGlob(path2, pattern) {
  let regex = pattern.replace(/[.+^${}()|[\]\\]/g, "\\$&").replace(/\*\*/g, "\xA7\xA7").replace(/\*/g, "[^/]*").replace(/§§\//g, "(.*/)?").replace(/§§/g, ".*");
  return new RegExp(`^${regex}$`).test(path2);
}

// src/indexer/chunk.ts
import { readFile } from "node:fs/promises";
import { relative as relative2 } from "node:path";

// src/engine.ts
var _engine = null;
var _engineId = "builtin";
async function getEngine() {
  if (_engine) return { engine: _engine, id: _engineId };
  try {
    const rustMod = await import("@docmd/engine-rust");
    if (typeof rustMod.isRustEngineAvailable === "function" && rustMod.isRustEngineAvailable()) {
      _engine = rustMod.createRustEngine();
      _engineId = "rust";
      return { engine: _engine, id: _engineId };
    }
  } catch {
  }
  try {
    const jsMod = await import("@docmd/engine-js");
    if (typeof jsMod.createJsEngine === "function") {
      _engine = jsMod.createJsEngine();
      _engineId = "js";
      return { engine: _engine, id: _engineId };
    }
  } catch {
  }
  _engine = createBuiltinEngine();
  _engineId = "builtin";
  return { engine: _engine, id: _engineId };
}
async function runTask(type, payload, throwOnError = false) {
  const { engine } = await getEngine();
  const result = await engine.run({ type, payload });
  if (result.success) return result.data ?? null;
  if (type.startsWith("search:")) {
    const fallbackResult = await runTaskWithFallback(type, payload);
    if (fallbackResult !== void 0) return fallbackResult;
  }
  if (throwOnError) throw new Error(result.error ?? `Task ${type} failed`);
  return null;
}
async function runTaskWithFallback(type, payload) {
  if (_engineId === "rust") {
    try {
      const jsMod = await import("@docmd/engine-js").catch(() => null);
      if (jsMod?.createJsEngine) {
        const jsEngine = jsMod.createJsEngine();
        const jsResult = await jsEngine.run({ type, payload });
        if (jsResult.success) return jsResult.data ?? null;
      }
    } catch {
    }
  }
  if (_engineId !== "builtin") {
    const builtin = createBuiltinEngine();
    const builtinResult = await builtin.run({ type, payload });
    if (builtinResult.success) return builtinResult.data ?? null;
  }
  return void 0;
}
async function getActiveEngineId() {
  await getEngine();
  return _engineId;
}
function createBuiltinEngine() {
  return {
    name: "builtin",
    supports(taskType) {
      return ["search:chunk", "search:quantize", "search:cosine"].includes(taskType);
    },
    async run(task) {
      const start = Date.now();
      try {
        let data;
        switch (task.type) {
          case "search:chunk":
            data = builtinChunk(task.payload);
            break;
          case "search:quantize":
            data = builtinQuantize(task.payload);
            break;
          case "search:cosine":
            data = builtinCosine(task.payload);
            break;
          default:
            return {
              success: false,
              error: `Built-in engine: unknown task '${task.type}'`,
              duration: Date.now() - start
            };
        }
        return { success: true, data, duration: Date.now() - start };
      } catch (err) {
        return {
          success: false,
          error: err?.message ?? String(err),
          duration: Date.now() - start
        };
      }
    }
  };
}
function builtinChunk({
  text,
  file,
  chunkSize = 256,
  chunkOverlap = 32
}) {
  const chunks = [];
  let currentHeading;
  let currentWords = [];
  let currentStart = 0;
  let bytePos = 0;
  for (const line of text.split("\n")) {
    const lineBytes = line.length + 1;
    if (/^#{1,6}\s/.test(line)) {
      if (currentWords.length > 0) {
        chunks.push({
          file,
          heading: currentHeading,
          text: currentWords.join(" "),
          range: [currentStart, bytePos]
        });
        currentWords = currentWords.slice(-chunkOverlap);
        currentStart = bytePos;
      }
      currentHeading = line.replace(/^#+\s*/, "").trim();
    } else {
      const words = line.split(/\s+/).filter(Boolean);
      currentWords.push(...words);
      if (currentWords.length >= chunkSize) {
        chunks.push({
          file,
          heading: currentHeading,
          text: currentWords.join(" "),
          range: [currentStart, bytePos + lineBytes]
        });
        currentWords = currentWords.slice(-chunkOverlap);
        currentStart = bytePos;
      }
    }
    bytePos += lineBytes;
  }
  if (currentWords.length > 0) {
    chunks.push({
      file,
      heading: currentHeading,
      text: currentWords.join(" "),
      range: [currentStart, bytePos]
    });
  }
  return chunks;
}
function builtinQuantize({
  vectors,
  dimensions = 384
}) {
  const quantized = [];
  const mins = [];
  const ranges = [];
  for (const vec of vectors) {
    const v = vec.length === dimensions ? vec : Array.from({ length: dimensions }, (_, i) => vec[i] ?? 0);
    const min = Math.min(...v);
    const max = Math.max(...v);
    const range = Math.abs(max - min) < 1e-10 ? 1 : max - min;
    const q = v.map((x) => Math.round((x - min) / range * 255 - 128));
    quantized.push(q);
    mins.push(min);
    ranges.push(range);
  }
  return { quantized, mins, ranges };
}
function builtinCosine({
  query,
  vectors,
  topK = 10
}) {
  const qNorm = Math.sqrt(query.reduce((s, v) => s + v * v, 0));
  if (qNorm < 1e-10) return [];
  const scores = vectors.map((vec, idx) => {
    const dot = query.reduce((s, v, i) => s + v * (vec[i] ?? 0), 0);
    const vNorm = Math.sqrt(vec.reduce((s, v) => s + v * v, 0));
    const sim = vNorm < 1e-10 ? 0 : dot / (qNorm * vNorm);
    return { index: idx, score: sim };
  });
  return scores.sort((a, b) => b.score - a.score).slice(0, topK);
}

// src/indexer/chunk.ts
async function chunkDocuments(files, rootDir, chunkSize, chunkOverlap) {
  const chunks = [];
  for (const file of files) {
    const content = await readFile(file, "utf-8");
    const rel = relative2(rootDir, file);
    const engineChunks = await runTask("search:chunk", {
      text: content,
      file: rel,
      chunkSize,
      chunkOverlap
    });
    if (engineChunks && engineChunks.length > 0) {
      chunks.push(...engineChunks);
    } else {
      chunks.push(...chunkText(content, rel, chunkSize, chunkOverlap));
    }
  }
  return chunks;
}
function chunkText(text, file, maxTokens, overlap) {
  const chunks = [];
  const sections = text.split(/^(#{1,6}\s+.+)$/m);
  let currentHeading;
  let buffer = "";
  let byteOffset = 0;
  for (const section of sections) {
    if (/^#{1,6}\s+/.test(section)) {
      if (buffer.trim()) {
        chunks.push(...splitBySize(buffer, file, currentHeading, byteOffset, maxTokens, overlap));
      }
      currentHeading = section.replace(/^#+\s+/, "").trim();
      byteOffset += Buffer.byteLength(buffer, "utf-8");
      buffer = "";
    } else {
      buffer += section;
    }
  }
  if (buffer.trim()) {
    chunks.push(...splitBySize(buffer, file, currentHeading, byteOffset, maxTokens, overlap));
  }
  return chunks;
}
function splitBySize(text, file, heading, startOffset, maxTokens, overlap) {
  const words = text.split(/\s+/).filter(Boolean);
  if (words.length <= maxTokens) {
    return [{
      file,
      heading,
      text: text.trim(),
      range: [startOffset, startOffset + Buffer.byteLength(text, "utf-8")]
    }];
  }
  const chunks = [];
  let i = 0;
  while (i < words.length) {
    const slice = words.slice(i, i + maxTokens).join(" ");
    chunks.push({
      file,
      heading,
      text: slice,
      range: [startOffset, startOffset + Buffer.byteLength(slice, "utf-8")]
    });
    i += maxTokens - overlap;
  }
  return chunks;
}

// src/config.ts
import { readFile as readFile2, writeFile, mkdir } from "node:fs/promises";
import { join as join2 } from "node:path";
import { homedir } from "node:os";
var AVAILABLE_MODELS = [
  {
    // ~23 MB quantized. English-only but very fast — good for EN-only docs.
    id: "Xenova/all-MiniLM-L6-v2",
    name: "MiniLM L6 v2 (English)",
    dimensions: 384,
    description: "Fastest. English-only. Best for single-language EN docs.",
    size: "~23 MB",
    recommended: true
  },
  {
    // ~118 MB quantized. 50+ languages, same 384-dim output as MiniLM.
    // Best all-round choice for multilingual docs.
    id: "Xenova/paraphrase-multilingual-MiniLM-L12-v2",
    name: "Multilingual MiniLM L12 v2",
    dimensions: 384,
    description: "Multilingual (50+ langs incl. Chinese, German). Recommended for i18n docs.",
    size: "~118 MB",
    recommended: false
  },
  {
    // ~118 MB quantized. 100+ languages via mE5 architecture.
    id: "Xenova/multilingual-e5-small",
    name: "Multilingual E5 Small",
    dimensions: 384,
    description: "Multilingual (100+ langs). Good retrieval quality.",
    size: "~118 MB",
    recommended: false
  },
  {
    // ~270 MB quantized. Highest quality multilingual option.
    id: "Xenova/paraphrase-multilingual-mpnet-base-v2",
    name: "Multilingual MPNet Base v2",
    dimensions: 768,
    description: "Best multilingual quality (50+ langs). Larger and slower.",
    size: "~270 MB",
    recommended: false
  }
];
function getModelProfile(modelId) {
  return AVAILABLE_MODELS.find((m) => m.id === modelId) ?? AVAILABLE_MODELS.find((m) => m.recommended);
}
function getDefaultModel() {
  return AVAILABLE_MODELS.find((m) => m.recommended);
}
var DEFAULT_EXCLUDE = [
  "**/node_modules/**",
  "**/dist/**",
  "**/build/**",
  "**/site/**",
  "**/.git/**",
  "**/.docmd-search/**",
  "**/.cache/**",
  "**/.next/**",
  "**/.nuxt/**",
  "**/coverage/**",
  "**/.svn/**",
  "**/.hg/**",
  "**/vendor/**"
];
var DEFAULT_CONFIG = {
  model: getDefaultModel().id,
  chunkSize: 256,
  chunkOverlap: 32,
  include: ["**/*.md", "**/*.txt", "**/*.html"],
  exclude: DEFAULT_EXCLUDE,
  outDir: ".docmd-search",
  incremental: true,
  topK: 10
};
var GLOBAL_DIR = join2(homedir(), ".docmd-search");
var GLOBAL_CONFIG_FILE = join2(GLOBAL_DIR, "config.json");
async function ensureGlobalDir() {
  await mkdir(GLOBAL_DIR, { recursive: true });
}
async function loadGlobalConfig() {
  try {
    const raw = await readFile2(GLOBAL_CONFIG_FILE, "utf-8");
    return JSON.parse(raw);
  } catch {
    return null;
  }
}
async function saveGlobalConfig(config) {
  await ensureGlobalDir();
  await writeFile(GLOBAL_CONFIG_FILE, JSON.stringify(config, null, 2), "utf-8");
}
async function isWizardCompleted() {
  const config = await loadGlobalConfig();
  return config?.wizardCompleted === true;
}
function getGlobalDir() {
  return GLOBAL_DIR;
}
async function loadProjectConfig(rootDir) {
  try {
    const configPath = join2(rootDir, ".docmd-search", "config.json");
    const raw = await readFile2(configPath, "utf-8");
    return JSON.parse(raw);
  } catch {
    return null;
  }
}
async function resolveConfig(rootDir, overrides) {
  const config = { ...DEFAULT_CONFIG };
  const global = await loadGlobalConfig();
  if (global?.model) {
    config.model = global.model;
  }
  if (rootDir) {
    const project = await loadProjectConfig(rootDir);
    if (project) {
      Object.assign(config, project);
    }
  }
  if (overrides) {
    Object.assign(config, overrides);
  }
  return config;
}

// src/model.ts
function checkPeerDeps() {
  const missing = [];
  try {
    import.meta.resolve?.("@huggingface/transformers");
  } catch {
    missing.push("@huggingface/transformers");
  }
  try {
    import.meta.resolve?.("onnxruntime-node");
  } catch {
    missing.push("onnxruntime-node");
  }
  return missing.length > 0 ? { missing } : null;
}
function formatMissingDepsMessage(missing) {
  const pkgs = missing.join(" ");
  return [
    "",
    "  \u26A0 Missing required dependencies for embedding:",
    "",
    `    npm install ${pkgs}`,
    "",
    "  These are optional peer dependencies \u2014 only needed when generating",
    "  embeddings. The search client (browser) does not need them.",
    ""
  ].join("\n");
}
async function quantizeToInt8(vectors) {
  const plain = vectors.map((v) => Array.from(v));
  const dimensions = plain[0]?.length ?? 384;
  const result = await runTask(
    "search:quantize",
    { vectors: plain, dimensions }
  );
  if (result?.quantized) {
    return result.quantized.map((q) => new Int8Array(q));
  }
  return vectors.map((vec) => {
    let min = Infinity;
    let max = -Infinity;
    for (let i = 0; i < vec.length; i++) {
      if (vec[i] < min) min = vec[i];
      if (vec[i] > max) max = vec[i];
    }
    const range = max - min || 1;
    const quantized = new Int8Array(vec.length);
    for (let i = 0; i < vec.length; i++) {
      quantized[i] = Math.round((vec[i] - min) / range * 255 - 128);
    }
    return quantized;
  });
}
async function getQuantizeEngine() {
  const id = await getActiveEngineId();
  switch (id) {
    case "rust":
      return "Rust engine";
    case "js":
      return "JS engine";
    default:
      return "built-in";
  }
}
var BATCH_SIZE = 128;
var MAX_CHARS_PER_TEXT = 1e3;
function createModelManager(modelId, onProgress) {
  const profile = getModelProfile(modelId);
  let pipeline = null;
  let loaded = false;
  const notify = (phase, message, progress) => {
    onProgress?.({ phase, message, progress });
  };
  return {
    profile,
    async load() {
      if (loaded) return;
      notify("loading", `Loading model: ${profile.name}...`);
      try {
        const { pipeline: createPipeline, env } = await import("@huggingface/transformers");
        const { homedir: homedir2 } = await import("os");
        const { join: join8 } = await import("path");
        env.cacheDir = join8(homedir2(), ".docmd-search", "models");
        const { detectGPUBackend: detectGPUBackend2, getExecutionProviders: getExecutionProviders2 } = await Promise.resolve().then(() => (init_gpu(), gpu_exports));
        const gpu = await detectGPUBackend2();
        if (gpu.hasGPU) {
          notify("loading", `GPU detected: ${gpu.name}`, void 0);
          const providers = getExecutionProviders2(gpu);
          if (env.backends.onnx) {
            env.backends.onnx.executionProviders = providers;
          }
        }
        let lastProgress = 0;
        const progressCallback = (data) => {
          if (data.status === "progress" && data.progress != null) {
            const pct = Math.round(data.progress);
            if (pct !== lastProgress) {
              lastProgress = pct;
              notify("downloading", `Downloading ${profile.name}...`, pct);
            }
          }
        };
        pipeline = await createPipeline("feature-extraction", profile.id, {
          progress_callback: progressCallback,
          dtype: "q8"
        });
        loaded = true;
        notify("ready", `Model ready: ${profile.name}${gpu.hasGPU ? ` (${gpu.name})` : ""}`);
      } catch (err) {
        const message = err?.message ?? String(err);
        if (message.includes("Cannot find module") || message.includes("MODULE_NOT_FOUND")) {
          throw new Error(
            `Missing dependency: @huggingface/transformers or onnxruntime-node.
Install with: npm install @huggingface/transformers onnxruntime-node`
          );
        }
        throw new Error(`Failed to load model "${profile.id}": ${message}`);
      }
    },
    isLoaded() {
      return loaded;
    },
    async embed(texts, onProgress2) {
      if (!loaded || !pipeline) {
        throw new Error("Model not loaded. Call load() first.");
      }
      const allEmbeddings = [];
      for (let i = 0; i < texts.length; i += BATCH_SIZE) {
        const batch = texts.slice(i, i + BATCH_SIZE);
        const truncated = batch.map(
          (t) => t.length > MAX_CHARS_PER_TEXT ? t.slice(0, MAX_CHARS_PER_TEXT) : t
        );
        const output = await pipeline(truncated, {
          pooling: "mean",
          normalize: true
        });
        for (let j = 0; j < batch.length; j++) {
          const embedding = new Float32Array(profile.dimensions);
          for (let k = 0; k < profile.dimensions; k++) {
            embedding[k] = output.data[j * profile.dimensions + k];
          }
          allEmbeddings.push(embedding);
        }
        onProgress2?.(allEmbeddings.length, texts.length);
      }
      return allEmbeddings;
    },
    async embedParallel(texts, onProgress2) {
      const { embedParallel: embedParallel2 } = await Promise.resolve().then(() => (init_parallel_embed(), parallel_embed_exports));
      return embedParallel2(texts, {
        modelId: profile.id,
        onProgress: onProgress2
      });
    },
    dispose() {
      pipeline = null;
      loaded = false;
    }
  };
}

// src/indexer/index.ts
init_index_io();

// src/git-aware.ts
import { stat as stat2 } from "node:fs/promises";
import { existsSync as existsSync2 } from "node:fs";
import { join as join5 } from "node:path";
import { execFile } from "node:child_process";
import { promisify } from "node:util";
var execFileAsync = promisify(execFile);
async function isGitRepo(dir) {
  const gitDir = join5(dir, ".git");
  return existsSync2(gitDir);
}
async function getChangedFilesFromGit(repoDir) {
  try {
    const { stdout } = await execFileAsync(
      "git",
      ["status", "--porcelain"],
      { cwd: repoDir }
    );
    const changed = /* @__PURE__ */ new Set();
    for (const line of stdout.split("\n")) {
      if (!line.trim()) continue;
      const filePath = line.slice(3).trim();
      if (line.startsWith(" D") || line.startsWith("D ")) {
        continue;
      }
      changed.add(filePath);
    }
    return changed;
  } catch (err) {
    return /* @__PURE__ */ new Set();
  }
}
async function getChangedFilesFromMtime(rootDir, filePaths, existingFiles) {
  const changed = [];
  for (const fp of filePaths) {
    const relative4 = fp.replace(rootDir + "/", "").replace(rootDir + "\\", "");
    const fileInfo = await stat2(fp);
    const existing = existingFiles[relative4];
    if (!existing || existing.mtime !== fileInfo.mtimeMs || existing.size !== fileInfo.size) {
      changed.push(fp);
    }
  }
  return changed;
}
async function detectChangedFiles(rootDir, filePaths, existingFiles) {
  if (await isGitRepo(rootDir)) {
    const gitChanged = await getChangedFilesFromGit(rootDir);
    if (gitChanged.size > 0) {
      const absolutePaths = Array.from(gitChanged).filter((relPath) => {
        const absPath = join5(rootDir, relPath);
        return filePaths.includes(absPath);
      }).map((relPath) => join5(rootDir, relPath));
      return absolutePaths;
    }
  }
  return getChangedFilesFromMtime(rootDir, filePaths, existingFiles);
}
async function getDetectionMethod(rootDir) {
  if (await isGitRepo(rootDir)) {
    return "git-aware";
  }
  return "mtime-based";
}

// src/model-cache.ts
var ModelCache = class {
  static instances = /* @__PURE__ */ new Map();
  static loadingPromises = /* @__PURE__ */ new Map();
  /**
   * Get or load a model by its ID.
   * If the model is already loaded, returns immediately.
   * If the model is currently loading, waits for the existing load to complete.
   * Otherwise, loads the model and caches it.
   */
  static async get(modelId, onProgress) {
    const cached = this.instances.get(modelId);
    if (cached && cached.isLoaded()) {
      return cached;
    }
    const existingLoad = this.loadingPromises.get(modelId);
    if (existingLoad) {
      await existingLoad;
      return this.instances.get(modelId);
    }
    const manager = createModelManager(modelId, onProgress);
    const loadPromise = manager.load();
    this.loadingPromises.set(modelId, loadPromise);
    try {
      await loadPromise;
      this.instances.set(modelId, manager);
    } finally {
      this.loadingPromises.delete(modelId);
    }
    return manager;
  }
  /**
   * Check if a model is currently loaded.
   */
  static has(modelId) {
    const instance = this.instances.get(modelId);
    return instance !== void 0 && instance.isLoaded();
  }
  /**
   * Dispose a specific model from cache.
   */
  static dispose(modelId) {
    const instance = this.instances.get(modelId);
    if (instance) {
      instance.dispose();
      this.instances.delete(modelId);
    }
  }
  /**
   * Dispose all cached models.
   */
  static disposeAll() {
    for (const instance of this.instances.values()) {
      instance.dispose();
    }
    this.instances.clear();
    this.loadingPromises.clear();
  }
  /**
   * Get list of currently loaded model IDs.
   */
  static getLoadedModels() {
    return Array.from(this.instances.keys());
  }
  /**
   * Get cache statistics.
   */
  static getStats() {
    const modelIds = this.getLoadedModels();
    return {
      loadedModels: modelIds.length,
      modelIds,
      memoryUsage: modelIds.length > 0 ? "~100-300 MB per model" : "0 MB"
    };
  }
};
async function getCachedModel(modelId, onProgress) {
  return ModelCache.get(modelId, onProgress);
}

// src/indexer/index.ts
var CHUNKS_PER_BATCH = 256;
async function indexDirectory(options, onProgress) {
  const {
    rootDir,
    include,
    exclude,
    chunkSize,
    chunkOverlap,
    outDir: outDirName,
    model: modelOverride,
    config: configOverride
  } = options;
  const config = await resolveConfig(rootDir, {
    ...configOverride,
    ...include ? { include } : {},
    ...exclude ? { exclude } : {},
    ...chunkSize != null ? { chunkSize } : {},
    ...chunkOverlap != null ? { chunkOverlap } : {},
    ...outDirName ? { outDir: outDirName } : {},
    ...modelOverride ? { model: modelOverride } : {}
  });
  const outDir = isAbsolute(config.outDir) ? config.outDir : join6(rootDir, config.outDir);
  const modelProfile = getModelProfile(config.model);
  const notify = (phase, current, total, file, message) => {
    onProgress?.({ phase, current, total, file, message });
  };
  notify("crawling", 0, 0, void 0, "Discovering files...");
  const filePaths = await crawl(rootDir, config.include, config.exclude);
  notify(
    "crawling",
    filePaths.length,
    filePaths.length,
    void 0,
    `Found ${filePaths.length} files`
  );
  const existingManifest = await loadManifest(outDir);
  let filesToIndex = filePaths;
  if (config.incremental && existingManifest) {
    if (existingManifest.status === "indexing") {
      notify("crawling", 0, 0, void 0, "Resuming interrupted indexing...");
    }
    const detectionMethod = await getDetectionMethod(rootDir);
    notify("crawling", 0, 0, void 0, `Detecting changes (${detectionMethod})...`);
    const changed = await detectChangedFiles(rootDir, filePaths, existingManifest.files);
    if (changed.length === 0 && existingManifest.status === "complete") {
      notify(
        "complete",
        filePaths.length,
        filePaths.length,
        void 0,
        "Index is up to date \u2014 no changes detected"
      );
      const { loadAllBatches: loadAllBatches2 } = await Promise.resolve().then(() => (init_index_io(), index_io_exports));
      return loadAllBatches2(outDir);
    }
    filesToIndex = changed.length > 0 ? changed : filePaths;
    if (changed.length > 0) {
      notify(
        "crawling",
        changed.length,
        filePaths.length,
        void 0,
        `Found ${changed.length} changed files (${detectionMethod})`
      );
    }
  }
  notify("chunking", 0, filesToIndex.length, void 0, "Chunking documents...");
  const chunks = await chunkDocuments(
    filesToIndex,
    rootDir,
    config.chunkSize,
    config.chunkOverlap
  );
  notify(
    "chunking",
    filesToIndex.length,
    filesToIndex.length,
    void 0,
    `Created ${chunks.length} chunks from ${filesToIndex.length} files`
  );
  const manifest = existingManifest ? { ...existingManifest, status: "indexing" } : createEmptyManifest(modelProfile.id, modelProfile.dimensions);
  manifest.model = modelProfile.id;
  manifest.dimensions = modelProfile.dimensions;
  manifest.totalFiles = filePaths.length;
  for (const fp of filesToIndex) {
    const rel = relative3(rootDir, fp);
    const fileInfo = await stat3(fp);
    manifest.files[rel] = {
      mtime: fileInfo.mtimeMs,
      size: fileInfo.size,
      batches: []
    };
  }
  await saveManifest(outDir, manifest);
  let modelManager = null;
  let allVectors = [];
  try {
    modelManager = await getCachedModel(config.model, (p) => {
      if (p.phase === "downloading") {
        notify(
          "downloading-model",
          p.progress ?? 0,
          100,
          void 0,
          p.message
        );
      }
    });
    notify("downloading-model", 0, 100, void 0, "Loading embedding model...");
    const allTexts = chunks.map((c) => c.text);
    notify("embedding", 0, chunks.length, void 0, "Generating embeddings...");
    const allFloatVectors = await modelManager.embed(allTexts, (current, total) => {
      notify(
        "embedding",
        current,
        total,
        chunks[current - 1]?.file,
        `Embedding ${current}/${total} chunks`
      );
    });
    notify("embedding", chunks.length, chunks.length, void 0, "Quantizing vectors...");
    const allInt8Vectors = await quantizeToInt8(allFloatVectors);
    let batchId = existingManifest?.batchCount ?? 0;
    const totalBatches = Math.ceil(chunks.length / CHUNKS_PER_BATCH);
    const compression = getCompressionType(chunks.length);
    for (let i = 0; i < chunks.length; i += CHUNKS_PER_BATCH) {
      const batchChunks = chunks.slice(i, i + CHUNKS_PER_BATCH);
      const batchVectors = allInt8Vectors.slice(i, i + CHUNKS_PER_BATCH);
      notify(
        "saving",
        batchId,
        totalBatches,
        void 0,
        `Saving batch ${batchId + 1}/${totalBatches}...`
      );
      await saveBatch(outDir, batchId, batchChunks, batchVectors, modelProfile.dimensions, compression);
      allVectors.push(...batchVectors);
      manifest.batchCount = batchId + 1;
      manifest.totalChunks = i + batchChunks.length;
      manifest.indexedFiles = new Set(
        chunks.slice(0, i + batchChunks.length).map((c) => c.file)
      ).size;
      await saveManifest(outDir, manifest);
      batchId++;
    }
  } catch (err) {
    if (chunks.length > 0 && allVectors.length === 0) {
      const zeroFloats = chunks.map(() => new Float32Array(modelProfile.dimensions));
      const zeroVectors = await quantizeToInt8(zeroFloats);
      await saveBatch(outDir, 0, chunks, zeroVectors, modelProfile.dimensions, "none");
      manifest.batchCount = 1;
      manifest.totalChunks = chunks.length;
      allVectors = zeroVectors;
    }
    manifest.status = "complete";
    await saveManifest(outDir, manifest);
    throw new Error(`Embedding failed: ${err?.message ?? String(err)}
Index saved with ${manifest.totalChunks} chunks (without embeddings).`);
  } finally {
    if (!options.keepModelLoaded) {
      modelManager?.dispose();
    }
  }
  manifest.status = "complete";
  manifest.indexedFiles = filePaths.length;
  await saveManifest(outDir, manifest);
  const navFiles = chunks.reduce((acc, chunk) => {
    if (!acc.has(chunk.file)) {
      acc.set(chunk.file, chunk.heading ?? "");
    }
    return acc;
  }, /* @__PURE__ */ new Map());
  const navEntries = [...navFiles.entries()].map(([path2, title]) => ({ path: path2, title }));
  await saveNavigation(outDir, buildNavTree(navEntries));
  notify(
    "complete",
    chunks.length,
    chunks.length,
    void 0,
    `Indexed ${chunks.length} chunks from ${filePaths.length} files`
  );
  return {
    version: 1,
    model: modelProfile.id,
    dimensions: modelProfile.dimensions,
    chunks,
    vectors: allVectors
  };
}

// src/index.ts
init_index_io();

// src/tui.ts
var A = {
  reset: "\x1B[0m",
  bold: "\x1B[1m",
  dim: "\x1B[2m",
  italic: "\x1B[3m",
  magenta: "\x1B[35m",
  green: "\x1B[32m",
  red: "\x1B[31m",
  yellow: "\x1B[33m",
  cyan: "\x1B[36m",
  white: "\x1B[37m",
  bgMagenta: "\x1B[45m",
  up: (n) => `\x1B[${n}A`,
  clear: "\x1B[2K",
  hide: "\x1B[?25l",
  show: "\x1B[?25h"
};
var BANNER = `
   ${A.magenta}     _                 _${A.reset}
   ${A.magenta}  __| |___ ___ _____ _| |${A.reset}
   ${A.magenta} | .  | . |  _|     | . |${A.reset}
   ${A.magenta} |____|___|___|_|_|_|___|${A.reset}
   ${A.dim} SEARCH${A.reset}
`;
function printBanner(version) {
  console.log(BANNER);
  if (version) {
    console.log(`   ${A.dim}v${version} \xB7 offline semantic search${A.reset}`);
    console.log("");
  }
}
async function printEngineStatus() {
  const id = await getActiveEngineId();
  switch (id) {
    case "rust":
      console.log(`   ${A.magenta}\u26A1${A.reset} ${A.bold}Rust engine${A.reset} ${A.dim}(accelerated)${A.reset}`);
      break;
    case "js":
      console.log(`   ${A.cyan}\u25C6${A.reset} ${A.bold}JS engine${A.reset} ${A.dim}(docmd)${A.reset}`);
      break;
    default:
      console.log(`   ${A.dim}\u25C7 built-in engine${A.reset}`);
  }
  console.log("");
}
function clearScreen() {
  process.stdout.write("\x1B[2J\x1B[3J\x1B[H");
}
function progressBar(current, total, width = 20) {
  if (total <= 0) return `${"\u2591".repeat(width)}`;
  const ratio = Math.min(current / total, 1);
  const filled = Math.round(width * ratio);
  const empty = width - filled;
  return `${"\u2588".repeat(filled)}${"\u2591".repeat(empty)}`;
}
function phaseLabel(phase) {
  switch (phase) {
    case "crawling":
      return "scanning";
    case "chunking":
      return "chunking";
    case "downloading-model":
      return "model   ";
    case "embedding":
      return "indexing ";
    case "saving":
      return "saving  ";
    case "complete":
      return "done    ";
    default:
      return phase;
  }
}
function createProgressDisplay(verbose = false) {
  let lastLines = 0;
  const startTime = Date.now();
  return {
    render(progress) {
      if (lastLines > 0) {
        process.stdout.write(A.up(lastLines) + "\r");
        for (let i = 0; i < lastLines; i++) {
          process.stdout.write(A.clear + "\n");
        }
        process.stdout.write(A.up(lastLines) + "\r");
      }
      const elapsed = ((Date.now() - startTime) / 1e3).toFixed(1);
      const pct = progress.total > 0 ? Math.round(progress.current / progress.total * 100) : 0;
      const bar = progressBar(progress.current, progress.total);
      const label = phaseLabel(progress.phase);
      let lines = [];
      if (progress.phase === "complete") {
        lines.push(`   ${A.green}\u2713${A.reset} ${progress.message ?? "Complete"}  ${A.dim}(${elapsed}s)${A.reset}`);
      } else {
        lines.push(`   ${A.magenta}${label}${A.reset}  ${bar}  ${A.bold}${pct}%${A.reset}  ${A.dim}${progress.current}/${progress.total}${A.reset}`);
        if (progress.file && verbose) {
          lines.push(`   ${A.dim}\u2192 ${progress.file}${A.reset}`);
        }
        if (progress.message && verbose) {
          lines.push(`   ${A.dim}${progress.message}${A.reset}`);
        }
        lines.push(`   ${A.dim}elapsed: ${elapsed}s${A.reset}`);
      }
      process.stdout.write(lines.join("\n") + "\n");
      lastLines = lines.length;
    },
    clear() {
      if (lastLines > 0) {
        process.stdout.write(A.up(lastLines) + "\r");
        for (let i = 0; i < lastLines; i++) {
          process.stdout.write(A.clear + "\n");
        }
        process.stdout.write(A.up(lastLines) + "\r");
        lastLines = 0;
      }
    }
  };
}
async function runSetupWizard() {
  const readline = await import("node:readline");
  console.log(`   ${A.bold}First-time setup${A.reset}`);
  console.log(`   ${A.dim}Select an embedding model for semantic search.${A.reset}`);
  console.log(`   ${A.dim}You can change this later with: docmd-search --settings${A.reset}`);
  console.log("");
  AVAILABLE_MODELS.forEach((model, i) => {
    const marker = model.recommended ? `${A.green}\u2605${A.reset}` : " ";
    const rec = model.recommended ? ` ${A.green}(recommended)${A.reset}` : "";
    console.log(`   ${marker} ${A.bold}${i + 1}.${A.reset} ${model.name}${rec}`);
    console.log(`      ${A.dim}${model.description}${A.reset}`);
    console.log(`      ${A.dim}${model.dimensions}d \xB7 ${model.size}${A.reset}`);
    console.log("");
  });
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  const answer = await new Promise((resolve2) => {
    rl.question(`   ${A.magenta}Pick a model${A.reset} [1-${AVAILABLE_MODELS.length}] (default: 1): `, resolve2);
  });
  rl.close();
  const choice = parseInt(answer.trim(), 10);
  const selectedIndex = choice >= 1 && choice <= AVAILABLE_MODELS.length ? choice - 1 : 0;
  const selected = AVAILABLE_MODELS[selectedIndex];
  console.log("");
  console.log(`   ${A.green}\u2713${A.reset} Selected: ${A.bold}${selected.name}${A.reset} ${A.dim}(${selected.size})${A.reset}`);
  console.log("");
  await saveGlobalConfig({
    model: selected.id,
    wizardCompleted: true
  });
  return selected;
}
async function runSearchTUI(searchFn, totalChunks) {
  const readline = await import("node:readline");
  console.log(`   ${A.dim}${totalChunks} chunks indexed \xB7 type to search \xB7 Ctrl+C to exit${A.reset}`);
  console.log("");
  if (!process.stdin.isTTY) {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
    rl.question(`   ${A.magenta}search${A.reset} \u276F `, (query2) => {
      const results = searchFn(query2, 10);
      printSearchResults(results, query2);
      rl.close();
    });
    return;
  }
  process.stdin.setRawMode(true);
  process.stdin.resume();
  process.stdin.setEncoding("utf-8");
  process.stdout.write(A.hide);
  let query = "";
  let lastResultLines = 0;
  const renderPrompt = () => {
    if (lastResultLines > 0) {
      process.stdout.write(A.up(lastResultLines + 1) + "\r");
      for (let i = 0; i <= lastResultLines; i++) {
        process.stdout.write(A.clear + "\n");
      }
      process.stdout.write(A.up(lastResultLines + 1) + "\r");
    }
    process.stdout.write(`${A.clear}   ${A.magenta}search${A.reset} \u276F ${query}
`);
    if (query.trim()) {
      const results = searchFn(query, 8);
      const lines = formatSearchResults(results, query);
      process.stdout.write(lines.join("\n") + "\n");
      lastResultLines = lines.length;
    } else {
      process.stdout.write(`   ${A.dim}start typing to search...${A.reset}
`);
      lastResultLines = 1;
    }
  };
  renderPrompt();
  process.stdin.on("data", (key) => {
    if (key === "") {
      process.stdout.write(A.show);
      process.stdout.write("\n");
      process.exit(0);
    }
    if (key === "\x7F" || key === "\b") {
      query = query.slice(0, -1);
    } else if (key === "\r" || key === "\n") {
      return;
    } else if (key.charCodeAt(0) >= 32 && key.charCodeAt(0) < 127) {
      query += key;
    } else {
      return;
    }
    renderPrompt();
  });
  await new Promise(() => {
  });
}
function formatSearchResults(results, query) {
  if (results.length === 0) {
    return [`   ${A.dim}no results for "${query}"${A.reset}`];
  }
  const lines = [""];
  for (let i = 0; i < results.length; i++) {
    const r = results[i];
    const pct = Math.round(r.score * 100);
    const scoreColor = pct > 60 ? A.green : pct > 30 ? A.yellow : A.dim;
    const heading = r.chunk.heading ? ` ${A.dim}\u203A${A.reset} ${r.chunk.heading}` : "";
    lines.push(
      `   ${A.dim}${String(i + 1).padStart(2)}.${A.reset} ${A.magenta}${r.chunk.file}${A.reset}${heading}  ${scoreColor}${pct}%${A.reset}`
    );
    const snippet = r.chunk.text.replace(/\s+/g, " ").trim().slice(0, 100);
    lines.push(`       ${A.dim}${snippet}${r.chunk.text.length > 100 ? "\u2026" : ""}${A.reset}`);
  }
  lines.push("");
  return lines;
}
function printSearchResults(results, query) {
  const lines = formatSearchResults(results, query);
  console.log(lines.join("\n"));
}
async function runSettingsTUI() {
  const readline = await import("node:readline");
  clearScreen();
  printBanner();
  console.log(`   ${A.bold}Settings${A.reset}`);
  console.log("");
  const globalConfig = await loadGlobalConfig();
  const currentModel = globalConfig?.model ? getModelProfile(globalConfig.model) : AVAILABLE_MODELS.find((m) => m.recommended);
  console.log(`   ${A.dim}Current model:${A.reset} ${A.bold}${currentModel.name}${A.reset} ${A.dim}(${currentModel.size})${A.reset}`);
  console.log("");
  console.log(`   ${A.bold}Available models:${A.reset}`);
  console.log("");
  AVAILABLE_MODELS.forEach((model, i) => {
    const isCurrent = model.id === currentModel.id;
    const marker = isCurrent ? `${A.green}\u25CF${A.reset}` : `${A.dim}\u25CB${A.reset}`;
    const label = isCurrent ? ` ${A.green}(active)${A.reset}` : "";
    const rec = model.recommended ? ` ${A.yellow}\u2605${A.reset}` : "";
    console.log(`   ${marker} ${A.bold}${i + 1}.${A.reset} ${model.name}${label}${rec}`);
    console.log(`      ${A.dim}${model.description}${A.reset}`);
    console.log(`      ${A.dim}${model.dimensions}d \xB7 ${model.size}${A.reset}`);
    console.log("");
  });
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  const answer = await new Promise((resolve2) => {
    rl.question(`   ${A.magenta}Switch model?${A.reset} [1-${AVAILABLE_MODELS.length}] (Enter to keep current): `, resolve2);
  });
  rl.close();
  const choice = parseInt(answer.trim(), 10);
  if (choice >= 1 && choice <= AVAILABLE_MODELS.length) {
    const selected = AVAILABLE_MODELS[choice - 1];
    if (selected.id === currentModel.id) {
      console.log(`   ${A.dim}Already using ${selected.name}${A.reset}`);
    } else {
      await saveGlobalConfig({
        model: selected.id,
        wizardCompleted: true
      });
      console.log("");
      console.log(`   ${A.green}\u2713${A.reset} Switched to: ${A.bold}${selected.name}${A.reset}`);
      console.log(`   ${A.dim}Model will be downloaded on next indexing run if needed.${A.reset}`);
    }
  } else {
    console.log(`   ${A.dim}Keeping ${currentModel.name}${A.reset}`);
  }
  console.log("");
}

// src/ui/launcher.ts
import { writeFile as writeFile3, mkdir as mkdir3, copyFile } from "node:fs/promises";
import { existsSync as existsSync3 } from "node:fs";
import { join as join7, resolve } from "node:path";
import { spawn, execFileSync } from "node:child_process";
var A2 = {
  reset: "\x1B[0m",
  bold: "\x1B[1m",
  dim: "\x1B[2m",
  magenta: "\x1B[35m",
  green: "\x1B[32m",
  red: "\x1B[31m",
  yellow: "\x1B[33m"
};
function generateDocmdConfig(rootDir, indexDir) {
  const absRootDir = resolve(rootDir).replace(/\\/g, "/");
  const absIndexDir = resolve(indexDir).replace(/\\/g, "/");
  const absCssFile = resolve(indexDir, "_ui", "custom.css").replace(/\\/g, "/");
  const safeSrc = JSON.stringify(absRootDir).slice(1, -1);
  const safeOut = JSON.stringify(absIndexDir + "/_site").slice(1, -1);
  const safeCss = JSON.stringify(absCssFile).slice(1, -1);
  const safeIdx = JSON.stringify(absIndexDir).slice(1, -1);
  return `// Auto-generated by docmd-search \u2014 do not edit manually.
// docmd-search has already built the semantic index; docmd just serves the UI.

export default {
  src: '${safeSrc}',
  out: '${safeOut}',
  title: 'docmd-search',
  description: 'Offline semantic search',

  theme: {
    customCss: ['${safeCss}'],
  },

  plugins: {
    // Semantic search with pre-built index \u2014 plugin-search skips indexing
    // and just reads from indexDir. No callback to docmd-search package.
    search: {
      semantic: true,
      showConfidence: true,
      indexDir: '${safeIdx}',
    },
  },

  // Disable features not needed for the search UI
  analytics: false,
  seo: false,
  sitemap: false,
};
`;
}
function generateCustomCSS() {
  return `/* docmd-search \u2014 custom theme (magenta palette) */
/* Injected via theme.customCss when launching docmd --ui */

:root {
  /* Primary colour overrides \u2014 docmd reads these tokens */
  --color-primary:       #d946ef;
  --color-primary-hover: #c026d3;
  --color-primary-light: rgba(217, 70, 239, 0.12);
  --color-accent:        #e879f9;

  /* docmd-search brand colours (used in custom components below) */
  --ds-magenta:  #d946ef;
  --ds-purple:   #a855f7;
  --ds-gradient: linear-gradient(135deg, var(--ds-magenta), var(--ds-purple));
}

/* \u2500\u2500 Badge \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
/* Shown in the corner to identify the docmd-search UI */
.ds-badge {
  position: fixed;
  bottom: 16px;
  right: 16px;
  background: var(--ds-gradient);
  color: #fff;
  padding: 5px 11px;
  border-radius: 20px;
  font-size: 11px;
  font-weight: 600;
  letter-spacing: 0.02em;
  opacity: 0.85;
  z-index: 9999;
  pointer-events: none;
  box-shadow: 0 2px 8px rgba(217, 70, 239, 0.3);
}

/* \u2500\u2500 Search input focus ring \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
.docmd-search-input:focus,
#docmd-search-input:focus {
  border-color: var(--ds-magenta) !important;
  box-shadow: 0 0 0 3px rgba(217, 70, 239, 0.15) !important;
  outline: none;
}

/* \u2500\u2500 Search result score badge \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
.search-score {
  background: var(--ds-gradient);
  color: #fff;
  padding: 2px 7px;
  border-radius: 10px;
  font-size: 10px;
  font-weight: 700;
  letter-spacing: 0.03em;
}

/* \u2500\u2500 Indexing progress banner \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
.ds-indexing-banner {
  background: rgba(217, 70, 239, 0.08);
  border: 1px solid rgba(217, 70, 239, 0.2);
  border-radius: 8px;
  padding: 12px 16px;
  margin: 16px 0;
  font-size: 13px;
  display: flex;
  align-items: center;
  gap: 10px;
  color: var(--ds-magenta);
}

.ds-indexing-spinner {
  width: 14px;
  height: 14px;
  border: 2px solid rgba(217, 70, 239, 0.25);
  border-top-color: var(--ds-magenta);
  border-radius: 50%;
  animation: ds-spin 0.75s linear infinite;
  flex-shrink: 0;
}

@keyframes ds-spin {
  to { transform: rotate(360deg); }
}

/* \u2500\u2500 Sidebar active link \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
.sidebar-link.active,
.nav-link.active {
  color: var(--ds-magenta) !important;
  border-left-color: var(--ds-magenta) !important;
}
`;
}
async function launchWebUI(options) {
  const { rootDir, indexDir, port = 0, open = true, verbose = false } = options;
  const stagingDir = join7(indexDir, "_ui");
  await mkdir3(stagingDir, { recursive: true });
  const cssContent = generateCustomCSS();
  await writeFile3(join7(stagingDir, "custom.css"), cssContent, "utf-8");
  const configContent = generateDocmdConfig(rootDir, indexDir);
  await writeFile3(join7(stagingDir, "docmd.config.json"), configContent, "utf-8");
  const indexNavFile = join7(indexDir, "navigation.json");
  const rootNavFile = join7(rootDir, "navigation.json");
  if (existsSync3(indexNavFile) && !existsSync3(rootNavFile)) {
    try {
      await copyFile(indexNavFile, rootNavFile);
      if (verbose) {
        console.log(`   ${A2.dim}navigation \u2192 ${rootNavFile}${A2.reset}`);
      }
    } catch {
    }
  }
  const indexMdPath = join7(rootDir, "index.md");
  if (!existsSync3(indexMdPath)) {
    const indexMdContent = `---
title: docmd-search
template: docmd-search
---

# Search Results

Use the search bar above to find content in this directory.

Results will appear here as you type.
`;
    try {
      await writeFile3(indexMdPath, indexMdContent, "utf-8");
      if (verbose) {
        console.log(`   ${A2.dim}index.md \u2192 ${indexMdPath}${A2.reset}`);
      }
    } catch {
    }
  }
  if (verbose) {
    console.log(`   ${A2.dim}staging \u2192 ${stagingDir}${A2.reset}`);
  }
  return new Promise((resolvePromise, reject) => {
    const resolvedBinary = resolveDocmdBinary(rootDir);
    let command;
    let args;
    if (resolvedBinary) {
      command = resolvedBinary;
      args = ["dev", "--config", join7(stagingDir, "docmd.config.json")];
      if (verbose) {
        console.log(`   ${A2.dim}using docmd at: ${resolvedBinary}${A2.reset}`);
      }
    } else {
      command = "npx";
      args = ["-y", "@docmd/core", "dev", "--config", join7(stagingDir, "docmd.config.json")];
      if (verbose) {
        console.log(`   ${A2.dim}falling back to: npx @docmd/core${A2.reset}`);
      }
    }
    if (port > 0) {
      args.push("--port", String(port));
    }
    const child = spawn(command, args, {
      cwd: rootDir,
      stdio: ["ignore", "pipe", "pipe"],
      env: {
        ...process.env,
        DOCMD_SEARCH_INDEX: indexDir
      }
    });
    let resolvedPort = port || 3e3;
    let resolved = false;
    if (child.stdout) {
      child.stdout.on("data", (data) => {
        const line = data.toString();
        if (verbose) process.stdout.write(line);
        const portMatch = line.match(/localhost:(\d+)/);
        if (portMatch && !resolved) {
          resolvedPort = parseInt(portMatch[1], 10);
          resolved = true;
          console.log(`   ${A2.magenta}\u25C6${A2.reset} Web UI ready`);
          console.log(`   ${A2.dim}http://localhost:${resolvedPort}${A2.reset}`);
          console.log("");
          if (open) {
            openBrowser(`http://localhost:${resolvedPort}`);
          }
          resolvePromise({
            port: resolvedPort,
            close: () => {
              child.kill("SIGTERM");
            }
          });
        }
      });
    }
    if (child.stderr && verbose) {
      child.stderr.on("data", (data) => process.stderr.write(data));
    }
    child.on("error", (err) => {
      if (err.message.includes("ENOENT") || err.message.includes("npx")) {
        console.log("");
        console.log(`   ${A2.red}\u2717${A2.reset} Could not launch web UI`);
        console.log("");
        console.log(`   ${A2.dim}Web UI requires @docmd/core. Install it with:${A2.reset}`);
        console.log(`     npm install -g @docmd/core`);
        console.log("");
        console.log(`   ${A2.dim}Or use the TUI search: docmd-search ./docs (without --ui)${A2.reset}`);
        console.log("");
      }
      reject(err);
    });
    child.on("exit", (code) => {
      if (!resolved) {
        if (code !== 0) {
          console.log("");
          console.log(`   ${A2.red}\u2717${A2.reset} docmd exited with code ${code}`);
          console.log("");
          console.log(`   ${A2.dim}Web UI requires @docmd/core. Install it with:${A2.reset}`);
          console.log(`     npm install -g @docmd/core`);
          console.log("");
          console.log(`   ${A2.dim}Or use the TUI search: docmd-search ./docs (without --ui)${A2.reset}`);
          console.log("");
        }
        reject(new Error(`docmd exited with code ${code}`));
      }
    });
    setTimeout(() => {
      if (!resolved) {
        resolved = true;
        resolvePromise({
          port: resolvedPort,
          close: () => {
            child.kill("SIGTERM");
          }
        });
      }
    }, 3e4);
  });
}
function openBrowser(url) {
  try {
    const cmd = process.platform === "darwin" ? "open" : process.platform === "win32" ? "start" : "xdg-open";
    spawn(cmd, [url], { detached: true, stdio: "ignore" }).unref();
  } catch {
  }
}
function resolveDocmdBinary(rootDir) {
  const absRootDir = resolve(rootDir);
  const searchRoots = [absRootDir, resolve(process.cwd())];
  const visited = /* @__PURE__ */ new Set();
  for (const start of searchRoots) {
    let dir = start;
    while (true) {
      if (visited.has(dir)) break;
      visited.add(dir);
      const candidate = join7(dir, "node_modules", ".bin", "docmd");
      if (existsSync3(candidate)) return candidate;
      const parent = join7(dir, "..");
      if (parent === dir) break;
      dir = parent;
    }
  }
  try {
    const whichCmd = process.platform === "win32" ? "where" : "which";
    const result = execFileSync(whichCmd, ["docmd"], { stdio: "pipe" }).toString().trim().split("\n")[0].trim();
    if (result && existsSync3(result)) return result;
  } catch {
  }
  return null;
}
export {
  AVAILABLE_MODELS,
  DEFAULT_CONFIG,
  buildNavTree,
  checkPeerDeps,
  clearScreen,
  compressVectors,
  createEmptyManifest,
  createModelManager,
  createProgressDisplay,
  createSearchIndex,
  decompressVectors,
  formatMissingDepsMessage,
  getActiveEngineId,
  getCompressionType,
  getDefaultModel,
  getEngine,
  getGlobalDir,
  getModelProfile,
  getQuantizeEngine,
  hasSearchableIndex,
  indexDirectory,
  isWizardCompleted,
  launchWebUI,
  loadAllBatches,
  loadBatch,
  loadGlobalConfig,
  loadManifest,
  loadProjectConfig,
  loadSearchIndex,
  printBanner,
  printEngineStatus,
  quantizeToInt8,
  resolveConfig,
  runSearchTUI,
  runSettingsTUI,
  runSetupWizard,
  runTask,
  saveBatch,
  saveGlobalConfig,
  saveManifest,
  saveNavigation,
  updateManifest
};
/**
 * --------------------------------------------------------------------
 * docmd-search : offline semantic search for docs, zero-config.
 *
 * @package     docmd-search (and ecosystem)
 * @website     https://docmd.io/search
 * @repository  https://github.com/docmd-io/docmd-search
 * @license     MIT
 * @copyright   Copyright (c) 2026-present docmd.io
 *
 * [docmd-source] - Please do not remove this header.
 * --------------------------------------------------------------------
 */
