// @docmd/deployer/dist/providers/github-pages.js
/**
 * --------------------------------------------------------------------
 * docmd : the zero-config documentation engine.
 *
 * @package     @docmd/deployer
 * @website     https://docmd.io
 * @repository  https://github.com/docmd-io/docmd
 * @license     MIT
 * @copyright   Copyright (c) 2025-present docmd.io
 *
 * [docmd-source] - Please do not remove this header.
 * --------------------------------------------------------------------
 */
export function generateGithubPages(ctx) {
    const buildCmd = ctx.configFile ? `docmd build --config ${ctx.configFile}` : 'docmd build';
    return `# ---------------------------------------------------
# GitHub Pages Deployment Workflow
# Generated by docmd v${ctx.version}
# Project: ${ctx.title}
# https://docmd.io/deployment/github-pages
# ---------------------------------------------------

name: Deploy to GitHub Pages

on:
  push:
    branches:
      - main

permissions:
  contents: read
  pages: write
  id-token: write

concurrency:
  group: "pages"
  cancel-in-progress: false

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v4

      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: "20"

      - name: Install dependencies
        run: npm install

      - name: Install docmd
        run: npm install -g @docmd/core@${ctx.version}

      - name: Build site
        run: ${buildCmd}

      - name: Upload Pages artifact
        uses: actions/upload-pages-artifact@v3
        with:
          path: ./${ctx.outDir}

  deploy:
    environment:
      name: github-pages
      url: \${{ steps.deployment.outputs.page_url }}
    runs-on: ubuntu-latest
    needs: build
    steps:
      - name: Deploy to GitHub Pages
        id: deployment
        uses: actions/deploy-pages@v4
`;
}
