// @awesome.me/webawesome/dist/webawesome.js
/*! Copyright 2026 Fonticons, Inc. - https://webawesome.com/license */
import "./chunks/chunk.D5YFE5NT.js";
import "./chunks/chunk.N2TXQSKF.js";
import {
  allDefined
} from "./chunks/chunk.X55YNZ3B.js";
import {
  serialize
} from "./chunks/chunk.LRYJ2M5H.js";
import "./chunks/chunk.KBS6YHTA.js";
import {
  discover,
  preventTurboFouce,
  startLoader,
  stopLoader
} from "./chunks/chunk.ZFSRFTCP.js";
import {
  getAnimationNames,
  getEasingNames
} from "./chunks/chunk.46TQDRI6.js";
import {
  registerTranslation
} from "./chunks/chunk.CDGKIW7Y.js";
import "./chunks/chunk.HK4J654O.js";
import {
  getDefaultIconFamily,
  registerIconLibrary,
  setDefaultIconFamily,
  unregisterIconLibrary
} from "./chunks/chunk.L2IYIH2C.js";
import {
  getIconFolder
} from "./chunks/chunk.D4VAJWKJ.js";
import {
  getBasePath,
  getIconPath,
  getKitCode,
  setBasePath,
  setIconPath,
  setKitCode
} from "./chunks/chunk.HGBRCPUS.js";
import "./chunks/chunk.XTA2JDH4.js";
import "./chunks/chunk.7VGCIHDG.js";
export {
  allDefined,
  discover,
  getAnimationNames,
  getBasePath,
  getDefaultIconFamily,
  getEasingNames,
  getIconFolder,
  getIconPath,
  getKitCode,
  preventTurboFouce,
  registerIconLibrary,
  registerTranslation,
  serialize,
  setBasePath,
  setDefaultIconFamily,
  setIconPath,
  setKitCode,
  startLoader,
  stopLoader,
  unregisterIconLibrary
};
