// @docmd/core/dist/commands/doctor.js
/**
 * --------------------------------------------------------------------
 * docmd doctor : pre-flight check command.
 *
 * Walks the user's config, checks every configured plugin and template
 * against the official registry, and prints a single status report. No
 * filesystem writes, no build side-effects — purely a diagnostic tool.
 *
 * Flags:
 *   --fix    Auto-install missing official plugins/templates.
 *   --json   Emit the report as JSON (for tooling).
 *   --config <path>  Path to docmd.config (defaults to the same search the
 *                    build commands use).
 *
 * Exit codes:
 *   0  — every check passed (or --fix resolved everything).
 *   1  — at least one problem remains (missing plugin, version mismatch,
 *         broken manifest).
 *   2  — usage error.
 * --------------------------------------------------------------------
 */
import path from 'node:path';
import fs from 'node:fs';
import { createRequire } from 'node:module';
import { fileURLToPath } from 'node:url';
import { TUI } from '@docmd/tui';
import { loadConfig } from '../utils/config-loader.js';
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
export async function runDoctor(opts) {
    // 1. Resolve the config.
    const configPath = opts.configPath ? path.resolve(opts.configPath) : null;
    let config = {};
    let configError = null;
    if (configPath) {
        try {
            const loaded = await loadConfig(configPath);
            config = loaded?.config || loaded || {};
        }
        catch (e) {
            configError = e.message;
        }
    }
    else {
        // loadConfig requires an explicit path; we don't have a built-in
        // "find a config" helper yet. The user can re-run with --config.
        configError = 'No config path provided. Use --config <path> to point at docmd.config.{json,ts,js,mjs}.';
    }
    // 2. Build the report.
    const report = {
        config: configPath,
        core: { name: '@docmd/core', version: 'unknown', status: 'ok' },
        plugins: [],
        engines: [],
        autoInstallCandidates: [],
        warnings: [],
        errors: configError ? [configError] : [],
    };
    // 3. Check @docmd/core version. We try a chain of strategies so the
    // doctor works in three real-world layouts:
    //   (a) a real user install — `node_modules/@docmd/core/package.json`
    //       is reachable via standard `createRequire` resolution.
    //   (b) monorepo dev with a built dist — the build output sits under
    //       `packages/core/dist/commands/`, and `import.meta.url` lets us
    //       derive the package root and read the in-tree `package.json`.
    //   (c) pnpm's symlinked `node_modules/@docmd/core` from inside the
    //       playground — walk up from cwd and try every ancestor.
    //
    // If all strategies fail, we genuinely don't have `@docmd/core` on the
    // resolve path; we report that rather than guessing.
    report.core = resolveCorePackage();
    if (report.core.status === 'missing') {
        report.errors.push('@docmd/core is not installed');
    }
    // 4. Check configured plugins.
    const configuredPlugins = (config.plugins || {});
    for (const [key, userOpts] of Object.entries(configuredPlugins)) {
        if (userOpts === false)
            continue;
        const pkgName = key.startsWith('@docmd/') ? key : `@docmd/plugin-${key}`;
        const installed = tryReadPackage(pkgName);
        if (installed) {
            report.plugins.push({
                name: key,
                status: 'ok',
                installedVersion: installed.version,
                fixable: false,
            });
        }
        else {
            report.plugins.push({
                name: key,
                status: 'missing',
                fixable: true,
                hint: `Run \`docmd add ${key}\` or add "${pkgName}" to your dependencies.`,
            });
            report.autoInstallCandidates.push(pkgName);
        }
    }
    // 5. Check the active template.
    if (config.theme?.template) {
        const tplName = String(config.theme.template).trim();
        const tplPkg = tplName.startsWith('@docmd/') ? tplName : `@docmd/template-${tplName}`;
        const installed = tryReadPackage(tplPkg);
        report.template = {
            requested: tplName,
            resolved: tplPkg,
            status: installed ? 'ok' : 'missing',
        };
        if (!installed) {
            report.autoInstallCandidates.push(tplPkg);
        }
    }
    // 6. Check engines. (The JS engine ships with core; the Rust engine is optional.)
    const requestedEngine = (config.engine || 'js').toLowerCase();
    report.engines.push({ name: 'js', status: 'ok' });
    if (requestedEngine === 'rust' || config.engines?.rust) {
        const rustInstalled = tryReadPackage('@docmd/engine-rust');
        report.engines.push({ name: 'rust', status: rustInstalled ? 'ok' : 'missing' });
    }
    // 7. Optional auto-fix.
    if (opts.fix && report.autoInstallCandidates.length > 0) {
        const { execSync } = await import('node:child_process');
        const cwd = process.cwd();
        const pkgManager = detectPackageManager(cwd);
        const cmd = `${pkgManager} add ${report.autoInstallCandidates.join(' ')}`;
        try {
            execSync(cmd, { stdio: 'inherit', cwd, timeout: 180000 });
        }
        catch (e) {
            report.errors.push(`Auto-install failed: ${e.message}`);
        }
    }
    // 8. Emit.
    if (opts.json) {
        console.log(JSON.stringify(report, null, 2));
    }
    else {
        printReport(report);
    }
    return report.errors.length > 0 ? 1 : 0;
}
function printReport(r) {
    console.log('');
    console.log('┌─ docmd pre-flight check');
    console.log('│');
    console.log(`│  ${r.core.name.padEnd(28)} ${r.core.version.padEnd(10)} ${r.core.status === 'ok' ? TUI.green('✓ installed') : TUI.red('✗ missing')}`);
    console.log('│');
    console.log('│  Configured plugins (' + r.plugins.length + ')');
    for (const p of r.plugins) {
        const sym = p.status === 'ok' ? TUI.green('✓') : p.status === 'missing' ? TUI.yellow('⚠') : TUI.dim('·');
        const ver = p.installedVersion ? ` (${p.installedVersion})` : '';
        console.log(`│    ${sym} ${p.name.padEnd(20)}${ver}${p.hint ? '  ' + TUI.dim(p.hint) : ''}`);
    }
    if (r.template) {
        const sym = r.template.status === 'ok' ? TUI.green('✓') : TUI.yellow('⚠');
        console.log(`│`);
        console.log(`│  Template`);
        console.log(`│    ${sym} ${r.template.resolved}${r.template.status === 'missing' ? '  ' + TUI.dim('not installed') : ''}`);
    }
    if (r.engines.length) {
        console.log('│');
        console.log('│  Engines');
        for (const e of r.engines) {
            const sym = e.status === 'ok' ? TUI.green('✓') : TUI.yellow('⚠');
            console.log(`│    ${sym} ${e.name}`);
        }
    }
    if (r.autoInstallCandidates.length) {
        console.log('│');
        console.log('│  ' + TUI.yellow(`Auto-install candidates (${r.autoInstallCandidates.length})`) + ' — run `docmd doctor --fix`');
        for (const c of r.autoInstallCandidates) {
            console.log(`│    • ${c}`);
        }
    }
    if (r.errors.length) {
        console.log('│');
        console.log('│  ' + TUI.red('Errors'));
        for (const e of r.errors)
            console.log(`│    ${e}`);
    }
    console.log('└──────────────────────────────────');
    console.log('');
}
/**
 * Resolve `@docmd/core`'s version with a chain of strategies. Returns
 * `{ name, version, status: 'ok' }` on success, or
 * `{ name, version: 'unknown', status: 'missing' }` if the package is
 * genuinely not on the resolve path.
 *
 * Strategy order:
 *   1. `createRequire(import.meta.url)` from the doctor file's URL —
 *      this works for a real user install where the doctor runs from
 *      `node_modules/@docmd/core/dist/commands/doctor.js`.
 *   2. Walk up from `process.cwd()` looking for a `node_modules/@docmd/core/package.json`
 *      — this covers pnpm's symlinked layout from the playground and
 *      yarn's hoisted layout in monorepos.
 *   3. Walk up from the doctor file's location looking for an in-tree
 *      `package.json` whose `name` is `@docmd/core` — this covers the
 *      monorepo dev case where the dist files are run directly without
 *      being installed into a `node_modules` tree.
 */
function resolveCorePackage() {
    const FAIL = { name: '@docmd/core', version: 'unknown', status: 'missing' };
    const SEARCH_NAMES = new Set(['@docmd/core', 'docmd']);
    // Strategy 1: createRequire from this file's URL.
    try {
        const req = createRequire(import.meta.url);
        const pkg = req('@docmd/core/package.json');
        return { name: pkg.name, version: pkg.version, status: 'ok' };
    }
    catch { /* fall through */ }
    // Strategy 2: walk up from cwd looking for node_modules/@docmd/core/package.json.
    for (let dir = process.cwd(); dir !== path.parse(dir).root; dir = path.dirname(dir)) {
        const candidate = path.join(dir, 'node_modules', '@docmd', 'core', 'package.json');
        if (fs.existsSync(candidate)) {
            try {
                const pkg = JSON.parse(fs.readFileSync(candidate, 'utf8'));
                if (SEARCH_NAMES.has(pkg.name)) {
                    return { name: pkg.name, version: pkg.version, status: 'ok' };
                }
            }
            catch { /* try next ancestor */ }
        }
    }
    // Strategy 3: walk up from the doctor file's location looking for an
    // in-tree package.json whose name is @docmd/core. This is the monorepo
    // dev case where the dist files are run directly without being installed
    // into a node_modules tree.
    let dir = __dirname;
    while (dir !== path.parse(dir).root) {
        const candidate = path.join(dir, 'package.json');
        if (fs.existsSync(candidate)) {
            try {
                const pkg = JSON.parse(fs.readFileSync(candidate, 'utf8'));
                if (SEARCH_NAMES.has(pkg.name)) {
                    return { name: pkg.name, version: pkg.version, status: 'ok' };
                }
            }
            catch { /* try next ancestor */ }
        }
        dir = path.dirname(dir);
    }
    return FAIL;
}
/**
 * Read a package's `package.json` from any of the three strategies above.
 * Used to detect configured plugins and engines.
 */
function tryReadPackage(name) {
    // Strategy 1: createRequire from this file's URL.
    try {
        const req = createRequire(import.meta.url);
        const pkg = req(`${name}/package.json`);
        return { name: pkg.name, version: pkg.version };
    }
    catch { /* fall through */ }
    // Strategy 2: walk up from cwd.
    for (let dir = process.cwd(); dir !== path.parse(dir).root; dir = path.dirname(dir)) {
        const candidate = path.join(dir, 'node_modules', ...name.split('/'), 'package.json');
        if (fs.existsSync(candidate)) {
            try {
                return JSON.parse(fs.readFileSync(candidate, 'utf8'));
            }
            catch { /* try next ancestor */ }
        }
    }
    // Strategy 3: walk up from the doctor file's location.
    let dir = __dirname;
    while (dir !== path.parse(dir).root) {
        const candidate = path.join(dir, 'node_modules', ...name.split('/'), 'package.json');
        if (fs.existsSync(candidate)) {
            try {
                return JSON.parse(fs.readFileSync(candidate, 'utf8'));
            }
            catch { /* try next ancestor */ }
        }
        dir = path.dirname(dir);
    }
    return null;
}
function detectPackageManager(cwd) {
    let dir = cwd;
    const { existsSync } = require('node:fs');
    const { join } = require('node:path');
    while (dir !== require('node:path').parse(dir).root) {
        if (existsSync(join(dir, 'pnpm-lock.yaml')))
            return 'pnpm';
        if (existsSync(join(dir, 'yarn.lock')))
            return 'yarn';
        if (existsSync(join(dir, 'bun.lockb')))
            return 'bun';
        if (existsSync(join(dir, 'package-lock.json')))
            return 'npm';
        dir = require('node:path').dirname(dir);
    }
    return 'npm';
}
